mlir/lib/Conversion/AIRRtToIpuPass.cpp

```diff
@@ -957,16 +1031,21 @@ struct AIRRtToIpuPass : public impl::AIRRtToIpuBase<AIRRtToIpuPass> {
     SmallVector<Type, 6> memrefTypes;
     SmallVector<Value, 6> memrefs;
     funcOp.walk([&](AIEX::IpuDmaMemcpyNdOp dma) {
-      // Value test_val = dyn_cast<Value>(dma.getMemref());
-      // auto test_val1 = dma.getMemref();
-      if (std::find(funcOp.getArguments().begin(), funcOp.getArguments().end(),
-                    dma.getMemref()) == funcOp.getArguments().end()) {
-        // push back if unique
-        if (std::find(memrefs.begin(), memrefs.end(), dma.getMemref()) ==
-            memrefs.end()) {
-          memrefs.push_back(dma.getMemref());
-          memrefTypes.push_back(dma.getMemref().getType());
-        }
+      Value memref = dma.getMemref();
+      auto args = funcOp.getArguments();
+      // if the memref is an arg, return
+      if (std::find(args.begin(), args.end(), memref) != args.end())
+        return;
+      // if the memref is the result of a cast of an arg, return
+      if (auto cast = dyn_cast_or_null<UnrealizedConversionCastOp>(memref.getDefiningOp()))
+        if (std::find(args.begin(), args.end(), cast.getOperand(0)) != args.end())
+          return;
+        else
+          memref = cast.getOperand(0);
+      // push back if unique
+      if (std::find(memrefs.begin(), memrefs.end(), dma.getMemref()) == memrefs.end()) {
+        memrefs.push_back(dma.getMemref());
+        memrefTypes.push_back(dma.getMemref().getType());
```

Comment:
This fixes a utility function that doesn't need to exist. Since we rely on the `func.func` to be in a certain form (e.g. with output parameters), the `func.func` to be lowered really should be in that form coming into the pass, instead of trying to patch it on the fly. We have at least [one test with broken IR](https://github.com/Xilinx/mlir-air/blob/main/mlir/test/Conversion/AIRRtToIpu/airrt_to_ipu.mlir#L387) depending on this fixup.